File: OpenResearch/ui/src/workspaceState.ts
export type Pane =
  | { kind: "home"; view: "experiments" | "files" | "artifacts" | "terminal" }
  | { kind: "experiment"; experimentId: string; view: "overview" | "terminal"; runId?: string }
  | { kind: "file"; path: string; source?: "repo" | "artifacts" | "abs"; sessionId?: string; ref?: string; line?: number; branchLabel?: string }
  | { kind: "code"; experimentId: string; branch: string; view: "files" | "changes" }
  | { kind: "plan"; sessionId: string; promptId: string }
  | { kind: "subagent"; sessionId: string; spawnPartId: string };

export interface TaskWorkspace {
  tabs: Pane[];
  active: Pane | null;
  previewKey: string | null;
  history: string[];
  expanded: Record<string, string[]>;
  scroll: Record<string, { top: number; left: number }>;
  sourceModes: Record<string, boolean>;
  filesView: "files" | "changes";
  scope: "agent" | "project";
  panelMax: boolean;
  /** Last React Flow viewport for the experiments tree, if it was opened. */
  treeViewport?: { x: number; y: number; zoom: number } | null;
}

export interface ProjectWorkspace {
  version: 1;
  lastTaskId: string | null;
  lastLocation: string | null;
  tasks: Record<string, TaskWorkspace>;
}

export function getTaskWorkspace(workspace: ProjectWorkspace | null | undefined, key: string): TaskWorkspace | undefined {
  return workspace && Object.hasOwn(workspace.tasks, key) ? workspace.tasks[key] : undefined;
}

export interface GlobalWorkspace {
  lastLocation: string | null;
  railOpen: boolean;
  panelWidth: number;
  experimentsView: "tree" | "table";
}

export const settingsTabs = ["settings", "harnesses", "projects", "compute", "instances", "environment", "git", "storage"] as const;
export type SettingsSection = typeof settingsTabs[number];
export const isSettingsSection = (value: unknown): value is SettingsSection =>
  settingsTabs.some((tab) => tab === value);

export const isRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === "object" && value !== null && !Array.isArray(value);
const nonempty = (value: unknown): value is string => typeof value === "string" && value.length > 0;
const optionalString = (value: unknown) => value == null || nonempty(value);

export function parsePane(value: unknown): Pane | undefined {
  if (!isRecord(value)) return;
  const only = (...fields: string[]) => Object.keys(value).every((field) => fields.includes(field));
  switch (value.kind) {
    case "home":
      if (only("kind", "view") && (value.view === "experiments" || value.view === "files" || value.view === "artifacts" || value.view === "terminal"))
        return { kind: "home", view: value.view };
      break;
    case "experiment":
      if (only("kind", "experimentId", "view", "runId") && nonempty(value.experimentId) && (value.view === "overview" || value.view === "terminal") && optionalString(value.runId))
        return { kind: "experiment", experimentId: value.experimentId, view: value.view, ...(nonempty(value.runId) ? { runId: value.runId } : {}) };
      break;
    case "file":
      if (only("kind", "path", "source", "sessionId", "ref", "line", "branchLabel") && nonempty(value.path) && optionalString(value.sessionId) && optionalString(value.ref) && optionalString(value.branchLabel)
        && (value.source == null || value.source === "repo" || value.source === "artifacts" || value.source === "abs")
        && (value.line == null || (typeof value.line === "number" && Number.isSafeInteger(value.line) && value.line > 0)))
        return { kind: "file", path: value.path, ...(value.source ? { source: value.source } : {}),
          ...(nonempty(value.sessionId) ? { sessionId: value.sessionId } : {}), ...(nonempty(value.ref) ? { ref: value.ref } : {}),
          ...(typeof value.line === "number" ? { line: value.line } : {}), ...(nonempty(value.branchLabel) ? { branchLabel: value.branchLabel } : {}) };
      break;
    case "code":
      if (only("kind", "experimentId", "branch", "view") && nonempty(value.experimentId) && nonempty(value.branch) && (value.view === "files" || value.view === "changes"))
        return { kind: "code", experimentId: value.experimentId, branch: value.branch, view: value.view };
      break;
    case "plan":
      if (only("kind", "sessionId", "promptId") && nonempty(value.sessionId) && nonempty(value.promptId)) return { kind: "plan", sessionId: value.sessionId, promptId: value.promptId };
      break;
    case "subagent":
      if (only("kind", "sessionId", "spawnPartId") && nonempty(value.sessionId) && nonempty(value.spawnPartId)) return { kind: "subagent", sessionId: value.sessionId, spawnPartId: value.spawnPartId };
  }
}

export interface Destination {
  kind: "home" | "resume" | "task" | "skills" | "settings";
  projectId?: string;
  sessionId?: string;
  section?: SettingsSection;
}

export function parseDestination(pathname: string): Destination | null {
  if (pathname === "/projects") return { kind: "home" };
  const parts = pathname.split("/");
  if (parts[0] !== "" || parts[1] !== "projects" || !parts[2]) return null;
  let projectId: string;
  let leaf: string;
  try { projectId = decodeURIComponent(parts[2]); leaf = decodeURIComponent(parts[4] ?? ""); } catch { return null; }
  const validId = (id: string) => id.length > 0 && id !== "." && id !== ".." && !/[\\/?#\u0000-\u001f\u007f-\u009f]/.test(id);
  if (!validId(projectId)) return null;
  if (parts.length === 3 || (parts.length === 4 && parts[3] === "")) return { kind: "resume", projectId };
  if (parts.length === 4 && parts[3] === "skills") return { kind: "skills", projectId };
  if (parts.length === 5 && parts[3] === "tasks" && validId(leaf)) return { kind: "task", projectId, ...(leaf === "new" ? {} : { sessionId: leaf }) };
  if (parts.length === 5 && parts[3] === "settings" && isSettingsSection(leaf)) return { kind: "settings", projectId, section: leaf };
  return null;
}

export function safeLocation(value: unknown): string | null {
  if (typeof value !== "string" || !value.startsWith("/") || value.startsWith("//") || /[\\#\u0000-\u001f\u007f-\u009f]/.test(value)) return null;
  const delimiter = value.indexOf("?");
  const pathname = delimiter === -1 ? value : value.slice(0, delimiter);
  const query = delimiter === -1 ? "" : value.slice(delimiter + 1);
  const destination = parseDestination(pathname);
  if (!destination || destination.kind === "resume") return null;
  const search = new URLSearchParams(query);
  if ([...search.keys()].some((key) => key !== "pane") || search.getAll("pane").length > 1) return null;
  if (search.has("pane")) {
    try { if (!parsePane(JSON.parse(search.get("pane") ?? ""))) return null; } catch { return null; }
  }
  return value;
}

export function taskLocation(projectId: string, sessionId: string | null, pane?: Pane | null): string {
  const path = `/projects/${encodeURIComponent(projectId)}/tasks/${sessionId ? encodeURIComponent(sessionId) : "new"}`;
  return pane ? `${path}?${new URLSearchParams({ pane: JSON.stringify(pane) })}` : path;
}

export const emptyProjectWorkspace = (): ProjectWorkspace => ({ version: 1, lastTaskId: null, lastLocation: null, tasks: {} });

// ponytail: one serialized writer per mounted workspace; add multi-window coordination only if needed.
export function createWorkspaceWriter<T>(save: (value: T, unloading: boolean) => Promise<unknown>, onError: (error: unknown) => void) {
  let pending: T | undefined;
  let failed: T | undefined;
  let running = false;
  let unloadPending = false;
  let timer: ReturnType<typeof setTimeout> | undefined;
  async function flush(unloading = false): Promise<void> {
    clearTimeout(timer);
    unloadPending ||= unloading;
    if (running || pending === undefined) return;
    running = true;
    const value = pending;
    pending = undefined;
    const keepalive = unloadPending;
    unloadPending = false;
    try { await save(value, keepalive); failed = undefined; }
    catch (error) { failed = value; onError(error); }
    finally { running = false; if (pending !== undefined) void flush(); }
  }
  return {
    queue(value: T, delay = 0) { pending = value; clearTimeout(timer); timer = setTimeout(() => void flush(), delay); },
    flush,
    retry() { if (!running && pending === undefined) pending = failed; return flush(); },
  };
}

File: OpenResearch/src/workspace_state.rs
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

use crate::error::{anyhow, Result};

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct WorkspaceState {
    pub version: u32,
    pub last_task_id: Option<String>,
    pub last_location: Option<String>,
    pub tasks: BTreeMap<String, TaskWorkspace>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct GlobalWorkspaceState {
    pub last_location: Option<String>,
    pub rail_open: bool,
    pub panel_width: f64,
    pub experiments_view: ExperimentsView,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct TaskWorkspace {
    tabs: Vec<Pane>,
    active: Option<Pane>,
    preview_key: Option<String>,
    history: Vec<String>,
    expanded: BTreeMap<String, Vec<String>>,
    scroll: BTreeMap<String, ScrollPosition>,
    source_modes: BTreeMap<String, bool>,
    files_view: FilesView,
    scope: Scope,
    panel_max: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    tree_viewport: Option<TreeViewport>,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(deny_unknown_fields)]
struct TreeViewport {
    x: f64,
    y: f64,
    zoom: f64,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(deny_unknown_fields)]
struct ScrollPosition {
    top: f64,
    left: f64,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(
    tag = "kind",
    rename_all = "camelCase",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum Pane {
    Home {
        view: HomeView,
    },
    Experiment {
        experiment_id: String,
        view: ExperimentView,
        #[serde(skip_serializing_if = "Option::is_none")]
        run_id: Option<String>,
    },
    File {
        path: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        source: Option<FileSource>,
        #[serde(skip_serializing_if = "Option::is_none")]
        session_id: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        r#ref: Option<String>,
        #[serde(skip_serializing_if = "Option::is_none")]
        line: Option<u64>,
        #[serde(skip_serializing_if = "Option::is_none")]
        branch_label: Option<String>,
    },
    Code {
        experiment_id: String,
        branch: String,
        view: FilesView,
    },
    Plan {
        session_id: String,
        prompt_id: String,
    },
    Subagent {
        session_id: String,
        spawn_part_id: String,
    },
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum HomeView {
    Experiments,
    Files,
    Artifacts,
    Terminal,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ExperimentView {
    Overview,
    Terminal,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum FileSource {
    Repo,
    Artifacts,
    Abs,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum FilesView {
    Files,
    Changes,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum Scope {
    Agent,
    Project,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum ExperimentsView {
    Tree,
    Table,
}

impl Pane {
    fn valid(&self) -> bool {
        let nonempty = |value: &String| !value.is_empty();
        match self {
            Self::Home { .. } => true,
            Self::Experiment {
                experiment_id,
                run_id,
                ..
            } => nonempty(experiment_id) && run_id.as_ref().is_none_or(nonempty),
            Self::File {
                path,
                session_id,
                r#ref,
                line,
                branch_label,
                ..
            } => {
                nonempty(path)
                    && session_id.as_ref().is_none_or(nonempty)
                    && r#ref.as_ref().is_none_or(nonempty)
                    && branch_label.as_ref().is_none_or(|label| !label.is_empty())
                    && line.is_none_or(|line| line > 0 && line <= 9_007_199_254_740_991)
            }
            Self::Code {
                experiment_id,
                branch,
                ..
            } => nonempty(experiment_id) && nonempty(branch),
            Self::Plan {
                session_id,
                prompt_id,
            } => nonempty(session_id) && nonempty(prompt_id),
            Self::Subagent {
                session_id,
                spawn_part_id,
            } => nonempty(session_id) && nonempty(spawn_part_id),
        }
    }
}

impl WorkspaceState {
    pub fn validate(&self) -> Result<()> {
        if self.version != 1
            || self.last_task_id.as_ref().is_some_and(|id| id.is_empty())
            || !self.last_location.as_deref().is_none_or(valid_location)
            || self.tasks.iter().any(|(id, task)| {
                id.is_empty()
                    || !task.tabs.iter().all(Pane::valid)
                    || !task.active.as_ref().is_none_or(Pane::valid)
                    || task.tree_viewport.as_ref().is_some_and(|viewport| {
                        !viewport.x.is_finite()
                            || !viewport.y.is_finite()
                            || !viewport.zoom.is_finite()
                            || viewport.zoom <= 0.0
                    })
                    || task.scroll.values().any(|position| {
                        !position.top.is_finite()
                            || !position.left.is_finite()
                            || position.top < 0.0
                            || position.left < 0.0
                    })
            })
        {
            return Err(anyhow!("invalid workspace state"));
        }
        Ok(())
    }

    pub fn from_stored(json: &str) -> Option<Self> {
        let state: Self = serde_json::from_str(json).ok()?;
        state.validate().ok()?;
        Some(state)
    }
}

impl GlobalWorkspaceState {
    pub fn validate(&self) -> Result<()> {
        if !self.last_location.as_deref().is_none_or(valid_location)
            || !self.panel_width.is_finite()
            || self.panel_width <= 0.0
        {
            return Err(anyhow!("invalid global workspace state"));
        }
        Ok(())
    }

    pub fn from_stored(json: &str) -> Option<Self> {
        let state: Self = serde_json::from_str(json).ok()?;
        state.validate().ok()?;
        Some(state)
    }
}

fn valid_location(location: &str) -> bool {
    if !location.starts_with('/')
        || location.starts_with("//")
        || location.contains(['\\', '#'])
        || location.chars().any(char::is_control)
    {
        return false;
    }
    let (path, query) = location.split_once('?').unwrap_or((location, ""));
    let parts: Vec<_> = path.split('/').collect();
    let valid_id = |id: &str| {
        urlencoding::decode(id).is_ok_and(|id| {
            !id.is_empty()
                && id != "."
                && id != ".."
                && !id.contains(['/', '\\', '?', '#'])
                && !id.chars().any(char::is_control)
        })
    };
    let valid_path = match parts.as_slice() {
        ["", "projects"] => true,
        ["", "projects", project, "skills"] => valid_id(project),
        ["", "projects", project, "tasks", task] => valid_id(project) && valid_id(task),
        ["", "projects", project, "settings", tab] => {
            valid_id(project)
                && urlencoding::decode(tab).is_ok_and(|tab| {
                    matches!(
                        tab.as_ref(),
                        "settings"
                            | "harnesses"
                            | "projects"
                            | "compute"
                            | "instances"
                            | "environment"
                            | "git"
                            | "storage"
                    )
                })
        }
        _ => false,
    };
    if !valid_path || query.is_empty() {
        return valid_path;
    }
    let Ok(url) = reqwest::Url::parse(&format!("http://localhost/?{query}")) else {
        return false;
    };
    let mut pairs = url.query_pairs();
    let Some((key, json)) = pairs.next() else {
        return true;
    };
    key == "pane"
        && pairs.next().is_none()
        && serde_json::from_str::<Pane>(&json)
            .ok()
            .is_some_and(|pane| pane.valid())
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn tree_viewport_is_optional_and_validated() {
        let mut value = json!({"version":1,"tasks":{"new":{
            "tabs":[],"active":null,"previewKey":null,"history":[],"expanded":{},
            "scroll":{},"sourceModes":{},"filesView":"files","scope":"project","panelMax":false
        }}});
        assert!(WorkspaceState::from_stored(&value.to_string()).is_some());
        for viewport in [json!(null), json!({"x":-24.0,"y":18.0,"zoom":1.4})] {
            value["tasks"]["new"]["treeViewport"] = viewport;
            let state = WorkspaceState::from_stored(&value.to_string()).unwrap();
            assert_eq!(
                WorkspaceState::from_stored(&serde_json::to_string(&state).unwrap()),
                Some(state)
            );
        }
        for viewport in [
            json!({"x":0,"y":0,"zoom":0}),
            json!({"x":0,"y":0,"zoom":-1}),
            json!({"x":null,"y":0,"zoom":1}),
        ] {
            value["tasks"]["new"]["treeViewport"] = viewport;
            assert!(WorkspaceState::from_stored(&value.to_string()).is_none());
        }
        value["tasks"]["new"]["treeViewport"] = json!({"x":0,"y":0,"zoom":1});
        let mut state: WorkspaceState = serde_json::from_value(value).unwrap();
        state
            .tasks
            .get_mut("new")
            .unwrap()
            .tree_viewport
            .as_mut()
            .unwrap()
            .x = f64::INFINITY;
        assert!(state.validate().is_err());
    }

    #[test]
    fn workspace_schema_and_resume_locations_are_validated() {
        let state = json!({"version":1,"lastLocation":"/projects/demo/tasks/new","tasks":{}});
        assert!(WorkspaceState::from_stored(&state.to_string()).is_some());
        for invalid in [
            json!({"version":2,"lastLocation":null,"tasks":{}}),
            json!({"version":1,"lastLocation":"https://example.com","tasks":{}}),
            json!({"version":1,"lastLocation":null,"tasks":{"new":{"tabs":[]}}}),
            json!({"version":1,"lastLocation":null,"tasks":{},"content":"not metadata"}),
        ] {
            assert!(WorkspaceState::from_stored(&invalid.to_string()).is_none());
        }
        for path in [
            "/",
            "/projects/demo",
            "//evil.test/projects",
            "/projects/../tasks/new",
            "/projects/demo/settings/unknown",
            "/projects/demo/tasks/new?pane=garbage",
            "/projects/demo/tasks/new#fragment",
        ] {
            assert!(!valid_location(path), "{path}");
        }
        let pane = json!({"kind":"file","path":"研究/figure +100%?draft#１.tex","source":"repo","line":9_007_199_254_740_991_u64});
        let location = format!(
            "/projects/demo/tasks/new?pane={}",
            urlencoding::encode(&pane.to_string())
        );
        assert!(valid_location(&location));
        assert!(valid_location("/projects/demo/settings/%67it"));
        assert!(valid_location(&format!(
            "/projects/demo/tasks/new?%70ane={}&",
            urlencoding::encode(&pane.to_string())
        )));
        assert!(valid_location("/projects/demo/tasks/new?pane=%7B%22kind%22%3A%22file%22%2C%22path%22%3A%22paper.tex%22%2C%22line%22%3Anull%7D"));
        let invalid_line =
            json!({"kind":"file","path":"paper.tex","line":9_007_199_254_740_992_u64});
        assert!(!serde_json::from_value::<Pane>(invalid_line)
            .unwrap()
            .valid());
        assert!(!Pane::File {
            path: "paper.tex".into(),
            source: None,
            session_id: None,
            r#ref: None,
            line: Some(0),
            branch_label: None,
        }
        .valid());
    }
}
